### DISCUSSION TITLE
Running out of memory when trying to yield a large amount of data from sql

### DISCUSSION THREAD

**Discussion by ddoan-code** [dagster-io/dagster#4669] [Q&A] [answered]
Hi there, I'm currently running into an issue where my code is causing my machine to run out of memory when reading a large SQL table using pandas.

My sql alchemy engine: 
`conn_str = f"mssql+pyodbc://{username}:{password}@{server}:{port}/{database}?driver={driver_2}"`
`engine = create_engine(conn_str)`
 `cnxn = engine.connect().execution_options(stream_results=True)`

I'm trying to read the table in chunks and then passing the dataframe on to other solids but this is causing a large spike in memory when the pipeline gets to a certain chunk. Changing the chunk size doesn't seem to help either.

`for idx, chunk in enumerate(pd.read_sql(sql, cnxn, chunksize=chunksize)):`
       `yield DynamicOutput(chunk, str(idx)` 
            
I'm not sure why it's causing the memory to spike because I'm using the FS IO manager so any help or insight would be greatly appreciated. Thanks

**Comment by sryza** [comment, marked as answer]
Hi @davidd-code - currently, Dagster won't release references to any of the dynamic outputs until the entire solid is complete.  That means that, at some point in your execution, all the chunks will be in memory at the same time.  We'd like to improve this, but aren't there yet.

**Comment by deoqc** [comment]
Hello dagsters,

Would be awesome if we could handle large data for 1 asset also w/ yield.

E.g.: have a large dataset that needs to be handled in chunks (but is not partitioned). I would yield the output of every chunk and the io_manager would be called for each chunk yielded.

**Reply by deoqc** [reply]
Just noticed that I can pass an iterator(*) and have iterable aware io manager.

The metadata on the runs will be bad, because the actual computation was pushed lazily to the io manager step (even thought it is not aware of it).

This is the best solution I could come up to the large/chunked data. Also believe this won't break in future...

(*) actually needs to be wrapped, if the iterator is passed directly, dagster won't pass it directly to io manager (will behave as if we had yielded, which seems bad)
